haystack/telemetry/_telemetry.py

```diff
@@ -144,18 +144,17 @@ def pipeline_running(pipeline: "Pipeline") -> Optional[Tuple[str, Dict[str, Any]
     pipeline._last_telemetry_sent = datetime.datetime.now()
 
     # Collect info about components
-    pipeline_description = pipeline.to_dict()
     components: Dict[str, List[Dict[str, Any]]] = defaultdict(list)
-    for component_name, component in pipeline_description["components"].items():
-        instance = pipeline.get_component(component_name)
+    for component_name, instance in pipeline.graph.nodes(data="instance"):
```

Comment:
Let's expose a `walk` method in `Pipeline` and call that instead. The method can walk the graph using BFS, yielding each component instance and its name

Reply:
I like that idea as an improvement of the code base. I would suggest we track that in a separate issue and PR because we want to get this P0 fix out quickly. We use that in quite a lot of places in the code base and would be good to improve them all then. WDYT?

Reply:
I'm inclined to disagree here - We should try to avoid committing quick hacks whenever possible. With the release branch-off approaching soon, I'd rather this issue was addressed in a correct manner now than punting it the future.

>We use that in quite a lot of places in the code base

What are you referring to here? I did a quick search for `.graph` and `.graph.nodes`, and I didn't come up any hits outside of the pipeline class and its tests.